### ISSUE TITLE
@UtilityClass does not create default no-args constructor

### ISSUE THREAD

**Issue by IDVsbruck** [rzwitserloot/lombok#2384]
**Describe the bug**
[Documentation of @UtilityClass](https://projectlombok.org/features/experimental/UtilityClass) shows an example where we can see that it creates a default no-args constructor, but when I test my class I see only one constructor with a parameter (parent object).

**To Reproduce**
Java 8+
JUnit, AssertJ
```
public class SomeClass {
    ...
    @UtilityClass
    public class SomeUtilityClass {
        public final String SOME_VALUE = "value";
    }
}
```
Test:
```
@Test
public void utilityClassTest() {
    final Constructor[] constructors = SomeClass.SomeUtilityClass.class.getDeclaredConstructors();
    for (Constructor constructor : constructors) {
        if (!Modifiers.isPrivate(constructor.getModifiers()) {
            Assert.fail();
        }
        constructor.setAccessible(true);
        Assert.assertEquals(0, constructor.getParameterCount());
        Assertions.assertThatThrownBy(constructor::newInstance).isInstanceOf(UnsupportedOperationException.class);
    }
}
```
**Expected behavior**
@UtilityClass class should contain one constructor SomeUtilityClass(){throw ...} instead current SomeUtilityClass(SomeClass someClass){throw ...}

**Version info (please complete the following information):**
 - Lombok version 1.18.12
 - Platform (javac or eclipse, and if so, what is the output of `javac -version` / the version listed in the _about..._ dialog of eclipse) - Java 11, Intellij IDEA

**Additional context**
No any additional context

**Comment by janrieke** [comment]
Could you check if it works when you compile outside of IntelliJ (e.g. with maven?
As there are automated test cases for this, I think it's a bug in the IntelliJ Lombok plugin.

**Comment by IDVsbruck** [comment]
Sure. I repeated this test with Maven 3.6.1, absolutely the same result - 1 parameter in the constructor.

**Comment by IDVsbruck** [comment]
If I set class annotated by @UtilityClass as **static** it works correctly.
But as I understand Lombok should make inner class with @UtilityClass static (and final).

**Comment by rzwitserloot** [comment]
@IDVsbruck Yeah, I think we can shortcut this issue down to: Actually, lombok ought to be making it static at which point the 1 param issue becomes irrelevant.

The one downside is that you can't put static classes inside non-static classes, but, how often are people gonna try that one? It'd have to look like:

```java
class Outer {
    class AnInnerClassThatIsNotStatic {
        @UtilityClass /* uh oh */
        class Helper {}
    }
}
```

In that case we should probably just error out and go: You just can't put utility classes in this location, it makes no sense to have a non-static one in the first place, and there's nothing lombok can do here to make it static. Besides, we can't make any elements inside this thing static either.

So, sounds like we should:

* Add 'static' if it isn't there and the annotated class is a class (and not an interface or enum or whatnot), and its enclosing type is a class (and not an interface or enum or whatnot). This is technically backwards incompatible, but given that it's [A] weird and [B] already doesn't work, I don't think it matters.
* Before adding static, check if the parent class is itself 'non-static' (which requires checking if it has a static modifier, and if it does not, if has an enclosing type and if it does, if that is a class and not an interface, enum, record, etc) – and then do absolutely no modifying of any code and generate an error message instead on that node.

and then look at @IDVsbruck and ask: Can you grab an edge release and see if you still have issues now, because I bet that'll solve them :)

Sounds good to all?

**Comment by IDVsbruck** [comment]
Thank you for your expand answer ...

I think anybody from us did not understand the issue (or my total misunderstanding).

No, my structure is different:
```
public class OuterNonStaticClass {
    @UtilityClass
    public class InnerNonStaticClass {
    }
}
```

Reading the documentation, I understand that @UtilityClass (in de-Lombok) makes InnerNonStaticClass as static and final, and adds private constructor without parameters. But in tests, I see a different picture ...

If I set InnerNonStaticClass as static, all functionality works correctly.

**Comment by rzwitserloot** [comment]
Cannot reproduce - but it sounds like this constructor is being created before lombok gets to run, as lombok definitely sets (and has since the introduction of UtilityClass) the static flag.

I need more details on how this is an actual issue ('I use reflection and that gives me a surprising result' doesn't count, I need an actual use case, and a recipe to reproduce this in a way that actually shows an error).

See also the test I just committed.
